storage/fanout.go

```diff
@@ -457,40 +446,76 @@ func NewMergeChunkSeriesSet(sets []ChunkSeriesSet, merger VerticalChunkSeriesMer
 		genericSets = append(genericSets, &genericChunkSeriesSetAdapter{s})
 
 	}
-	return &chunkSeriesSetAdapter{newGenericMergeSeriesSet(genericSets, nil, (&chunkSeriesMergerAdapter{VerticalChunkSeriesMergerFunc: merger}).Merge)}
+	return &chunkSeriesSetAdapter{newGenericMergeSeriesSet(genericSets, (&chunkSeriesMergerAdapter{VerticalChunkSeriesMergerFunc: merger}).Merge)}
 }
 
 // newGenericMergeSeriesSet returns a new genericSeriesSet that merges (and deduplicates)
 // series returned by the chkQuerierSeries series sets when iterating.
 // Each chkQuerierSeries series set must return its series in labels order, otherwise
 // merged series set will be incorrect.
-// Argument 'querier' is optional and can be nil. Pass Querier if you want to retry query in case of failing series set.
 // Overlapped situations are merged using provided mergeFunc.
-func newGenericMergeSeriesSet(sets []genericSeriesSet, querier *mergeGenericQuerier, mergeFunc genericSeriesMergeFunc) genericSeriesSet {
+func newGenericMergeSeriesSet(sets []genericSeriesSet, mergeFunc genericSeriesMergeFunc) genericSeriesSet {
 	if len(sets) == 1 {
 		return sets[0]
 	}
 
+	mset := &genericMergeSeriesSet{
+		mergeFunc: mergeFunc,
+		sets:      sets,
+	}
 	// Sets need to be pre-advanced, so we can introspect the label of the
 	// series under the cursor.
-	var h genericSeriesSetHeap
-	for _, set := range sets {
-		if set == nil {
-			continue
-		}
-		if set.Next() {
-			heap.Push(&h, set)
+	mset.advance()
+	return mset
+}
+
+func (c *genericMergeSeriesSet) advance() {
+	if !c.advanced {
```

Comment:
This has unnecessary indent.. (: Can you just return if `c.advanced?`

Reply:
This is a code smell as well. Why would you `advance` if it was advanced before? It's better if caller will make sure this will not happen, otherwise we can end up in unnecessary  invoking `advance` thousand times `just in case`. This creates confusion and is extremely hard to follow. I would fix this TBH before merging.